aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/CachedValue.kt

```
@@ -29,34 +33,58 @@ internal class CachedValue<T> (
     private var value: ExpiringValue<T>? = null,
     private val bufferTime: Duration = Duration.ZERO,
     private val clock: Clock = Clock.System,
-) {
+) : Closeable {
     constructor(value: T, expiresAt: Instant, bufferTime: Duration = Duration.ZERO, clock: Clock = Clock.System) : this(ExpiringValue(value, expiresAt), bufferTime, clock)
-    private val mu = Mutex()
+
+    private val gate = Semaphore(1)
+    private val _ref: AtomicRef<ExpiringValue<T>?> = atomic(value)
+    private val ref: ExpiringValue<T>?
+        get() = _ref.value
+    private val closed = atomic(false)
 
     /**
-     * Check if the value is expired or not as compared to the time [now]
+     * Check if the value is expired or not as compared to the time now
      */
-    suspend fun isExpired(): Boolean = mu.withLock { isExpiredUnlocked() }
-
-    private fun isExpiredUnlocked(): Boolean {
-        val curr = value ?: return true
-        return clock.now() >= (curr.expiresAt - bufferTime)
+    fun isExpired(): Boolean {
+        check(!closed.value) { "value is closed" }
+        return ref?.let { isExpired(it) } ?: true
     }
 
+    private fun isExpired(value: ExpiringValue<T>): Boolean = clock.now() >= (value.expiresAt - bufferTime)
+
     /**
      * Get the value if it has not expired yet. Returns null if the value has expired
      */
-    suspend fun get(): T? = mu.withLock {
-        if (!isExpiredUnlocked()) return value!!.value else null
+    fun get(): T? {
+        check(!closed.value) { "value is closed" }
+
+        if (ref == null || isExpired(ref!!)) {
```

**correctness**: This is doing 3 atomic loads of a `ref`. Typically with atomics you'll want to load and store it into a local variable.

e.g.

```kotlin
check(!closed) { ... }
val curr = ref ?: return null
return isExpired(curr)
```

This ensures the value isn't updated from a different thread/coroutine in middle of a single function call.

---

Moreover since this doesn't modify the value should this throw an exception or just return `null`? 🤔 

```kotlin
fun get(): T? {
    val curr = ref
    if (curr == null || !isClosed || isExpired(curr)) return null
    return curr.takeIf { !isClosed && it != null && !isExpired(it) }
}
```


Summon @ianbotsf.

---

I would argue that it calling `get` on a closed instance should throw an exception, not return `null`.

```kotlin
fun get(): T? {
    check(!closed.value) { "value is closed" }
    return ref?.takeIf { !isExpired(it) }
}
```